sdks/java/io/kinesis/src/test/java/org/apache/beam/sdk/io/kinesis/KinesisIOIT.java

```diff
@@ -0,0 +1,126 @@
+  public static final int RECORDS_TOTAL = 1000;
+  public static final int SHARDS_TOTAL = 2;
+
+  @Rule public final transient TestPipeline p = TestPipeline.create();
+  @Rule public final transient TestPipeline p2 = TestPipeline.create();
+
+  private static KinesisTestOptions options;
+
+  @BeforeClass
+  public static void setup() {
+    PipelineOptionsFactory.register(KinesisTestOptions.class);
+    options = TestPipeline.testingPipelineOptions().as(KinesisTestOptions.class);
+  }
+
+  @Test
+  public void testWriteThenRead() throws Exception {
+    Instant now = Instant.now();
+    List<byte[]> data = prepareData();
+
+    // Write data into stream
+    p.apply(Create.of(data))
+        .apply(
+            KinesisIO.write()
+                .withStreamName(options.getAwsKinesisStream())
+                .withPartitioner(new RandomPartitioner())
+                .withAWSClientsProvider(
+                    options.getAwsAccessKey(),
+                    options.getAwsSecretKey(),
+                    Regions.fromName(options.getAwsKinesisRegion())));
+    p.run().waitUntilFinish();
+
+    // Read new data from stream that was just written before
+    PCollection<byte[]> output =
+        p2.apply(
+                KinesisIO.read()
+                    .withStreamName(options.getAwsKinesisStream())
+                    .withAWSClientsProvider(
+                        options.getAwsAccessKey(),
+                        options.getAwsSecretKey(),
+                        Regions.fromName(options.getAwsKinesisRegion()))
+                    .withMaxNumRecords(data.size())
+                    // to prevent endless running in case of error
+                    .withMaxReadTime(Duration.standardMinutes(5))
+                    .withInitialPositionInStream(InitialPositionInStream.AT_TIMESTAMP)
+                    .withInitialTimestampInStream(now))
+            .apply(
+                ParDo.of(
+                    new DoFn<KinesisRecord, byte[]>() {
+
+                      @ProcessElement
+                      public void processElement(ProcessContext c) {
+                        KinesisRecord record = c.element();
+                        byte[] data = record.getData().array();
+                        c.output(data);
+                      }
+                    }));
+    PAssert.that(output).containsInAnyOrder(data);
+    p2.run().waitUntilFinish();
+  }
+
+  private List<byte[]> prepareData() {
+    List<byte[]> data = newArrayList();
+    for (int i = 0; i < RECORDS_TOTAL; i++) {
+      data.add(String.valueOf(i).getBytes());
+    }
+    return data;
+  }
+
+  private static final class RandomPartitioner implements KinesisPartitioner {
+    @Override
+    public String getPartitionKey(byte[] value) {
+      Random rand = new Random();
+      int n = rand.nextInt(SHARDS_TOTAL) + 1;
+      return String.valueOf(n);
+    }
+
+    @Override
+    public String getExplicitHashKey(byte[] value) {
+      return null;
```

Comment:
is there a risk of NPE?  `getExplicitHashKey()` is fed to ` producer.addUserRecord()` with no null check. If `producer.addUserRecord()` is tolerant to null hashkey then it is ok, if not, need to protect against NPE

Reply:
ok